File: iOS/Crisis Connect/Features/Messaging/SfuRoomClient.swift
//
//  SfuRoomClient.swift
//  Crisis Connect
//
//  SFU room roster over Firestore — the coordination plane the SFU itself doesn't provide.
//  Mirrors the web `sfu-room.ts` and Android's SfuRoomClient: `sfuRooms/{roomId}/participants/{uid}`
//  holds each member's SFU sessionId + published track names so everyone can pull everyone else.
//  Carries NO media and NO plaintext keys — MLS handshake bytes ride the sibling `mlsMessages`
//  subcollection as opaque blobs. Every write carries `expireAt` so a Firestore TTL policy sweeps
//  orphans (a crashed peer's roster doc, the handshake blobs); live calls keep extending it.
//

import FirebaseFirestore
import Foundation

enum SfuRoomScopeType: String {
    case agency, hierarchy
}

enum SfuRoomAuthorizationError: Error {
    case invalidBinding
    case roomBindingMismatch
}

/// Immutable authorization record for one accepted call. The room UUID is only a routing key;
/// Firestore rules and every client bind it to this exact scope, call id, and participant pair.
struct SfuRoomBinding: Equatable {
    let scopeType: SfuRoomScopeType
    let channelId: String
    let callId: String
    let participants: [String]

    init(
        scopeType: SfuRoomScopeType,
        channelId: String,
        callId: String,
        selfUid: String,
        peerUid: String
    ) throws {
        let channel = channelId.trimmingCharacters(in: .whitespacesAndNewlines)
        let call = callId.trimmingCharacters(in: .whitespacesAndNewlines)
        let me = selfUid.trimmingCharacters(in: .whitespacesAndNewlines)
        let peer = peerUid.trimmingCharacters(in: .whitespacesAndNewlines)
        guard !channel.isEmpty, channel.count <= 256,
              UUID(uuidString: call) != nil,
              !me.isEmpty, me.count <= 128,
              !peer.isEmpty, peer.count <= 128,
              me != peer else {
            throw SfuRoomAuthorizationError.invalidBinding
        }
        self.scopeType = scopeType
        self.channelId = channel
        self.callId = call
        self.participants = [me, peer].sorted()
    }

    var documentFields: [String: Any] {
        [
            "version": 2,
            "scopeType": scopeType.rawValue,
            "channelId": channelId,
            "callId": callId,
            "participants": participants
        ]
    }

    func matches(_ snapshot: DocumentSnapshot) -> Bool {
        guard (snapshot.get("version") as? NSNumber)?.intValue == 2,
              snapshot.get("scopeType") as? String == scopeType.rawValue,
              snapshot.get("channelId") as? String == channelId,
              snapshot.get("callId") as? String == callId,
              snapshot.get("participants") as? [String] == participants,
              let creator = snapshot.get("mlsCreator") as? String,
              participants.contains(creator) else {
            return false
        }
        return true
    }
}

final class SfuRoomClient: @unchecked Sendable {
    private let roomId: String
    private let selfUid: String
    private let binding: SfuRoomBinding
    private let protocolVersion: SfuProtocolVersion
    private let db: Firestore

    var enforcesBoundIdentity: Bool { true }

    // 12h: comfortably longer than any real authority call, refreshed by heartbeat while live.
    private static let roomTtl: TimeInterval = 12 * 60 * 60

    init(
        roomId: String,
        selfUid: String,
        binding: SfuRoomBinding,
        protocolVersion: SfuProtocolVersion = .secure,
        firestore: Firestore = .firestore()
    ) {
        self.roomId = roomId
        self.selfUid = selfUid
        self.binding = binding
        self.protocolVersion = protocolVersion
        self.db = firestore
    }

    private var participants: CollectionReference {
        db.collection(roomCollection).document(roomId).collection("participants")
    }

    private var mls: CollectionReference {
        db.collection(roomCollection).document(roomId).collection("mlsMessages")
    }

    private var roomCollection: String {
        "sfuRoomsV2"
    }

    private func expireAt() -> Timestamp {
        Timestamp(date: Date().addingTimeInterval(Self.roomTtl))
    }

    /// Announce (or update) my SFU identity so other members can pull my tracks.
    func publishSelf(
        name: String,
        photoUrl: String?,
        cameraOn: Bool,
        muted: Bool,
        sessionId: String,
        tracks: [SfuPublishedTrack],
        onError: @escaping (Error) -> Void = { _ in }
    ) {
        participants.document(selfUid).setData([
            "uid": selfUid,
            "name": name,
            "photoUrl": photoUrl as Any,
            "cameraOn": cameraOn,
            "muted": muted,
            "sessionId": sessionId,
            "tracks": tracks.map {
                ["trackName": $0.trackName, "kind": $0.kind, "source": $0.source]
            },
            "updatedAt": FieldValue.serverTimestamp(),
            "expireAt": expireAt()
        ], completion: { error in
            if let error { onError(error) }
        })
    }

    /// Remove myself from the room roster on hang-up (best-effort).
    func leave() {
        participants.document(selfUid).delete()
    }

    /// Liveness heartbeat: refresh my roster entry so peers can tell a live member from a
    /// crashed one (an ungraceful exit never deletes the doc — only staleness reveals it).
    func heartbeat(onError: @escaping (Error) -> Void = { _ in }) {
        participants.document(selfUid).updateData([
            "updatedAt": FieldValue.serverTimestamp(),
            "expireAt": expireAt()
        ], completion: { error in
            if let error { onError(error) }
        })
    }

    /// Atomically claim the MLS group-creator role on the room doc. Exactly one member gets
    /// `true` (race-free); everyone else joins the group the creator makes.
    func claimMlsCreator() async throws -> Bool {
        let reference = db.collection(roomCollection).document(roomId)
        let selfUid = selfUid
        let binding = binding
        let protocolVersion = protocolVersion
        let expire = expireAt()
        guard protocolVersion == .secure, binding.participants.contains(selfUid) else {
            throw SfuRoomAuthorizationError.invalidBinding
        }
        let result = try await db.runTransaction { transaction, errorPointer -> Any? in
            do {
                let snapshot = try transaction.getDocument(reference)
                if snapshot.exists {
                    if !binding.matches(snapshot) {
                        errorPointer?.pointee = SfuRoomAuthorizationError.roomBindingMismatch as NSError
                        return nil
                    }
                    if let creator = snapshot.get("mlsCreator") as? String, !creator.isEmpty {
                        return creator == selfUid
                    }
                    errorPointer?.pointee = SfuRoomAuthorizationError.roomBindingMismatch as NSError
                    return nil
                }
                var fields = binding.documentFields
                fields["mlsCreator"] = selfUid
                fields["createdAt"] = FieldValue.serverTimestamp()
                fields["expireAt"] = expire
                transaction.setData(fields, forDocument: reference)
                return true
            } catch {
                errorPointer?.pointee = error as NSError
                return nil
            }
        }
        return result as? Bool ?? false
    }

    /// Subscribe to the room roster; delivers everyone except me.
    func listenRoster(
        onRoster: @escaping ([SfuRemoteParticipant]) -> Void,
        onError: @escaping (Error) -> Void = { _ in }
    ) -> ListenerRegistration {
        participants.addSnapshotListener { [selfUid] snapshot, error in
            if let error { onError(error); return }
            guard let snapshot else { return }
            let remotes: [SfuRemoteParticipant] = snapshot.documents.compactMap { document in
                guard document.documentID != selfUid,
                      let sessionId = document.get("sessionId") as? String,
                      let rawTracks = document.get("tracks") as? [[String: Any]] else {
                    return nil
                }
                let tracks: [SfuPublishedTrack] = rawTracks.compactMap { raw in
                    guard let trackName = raw["trackName"] as? String else { return nil }
                    return SfuPublishedTrack(
                        trackName: trackName,
                        kind: raw["kind"] as? String ?? "audio",
                        source: raw["source"] as? String ?? "mic"
                    )
                }
                return SfuRemoteParticipant(
                    uid: document.documentID,
                    sessionId: sessionId,
                    tracks: tracks,
                    updatedAt: (document.get("updatedAt") as? Timestamp)?.dateValue()
                )
            }
            onRoster(remotes)
        }
    }

    // MARK: - MLS handshake relay (opaque blobs; Firestore never sees keys/plaintext)

    /// Broadcast one opaque MLS handshake message (codec JSON) to the room.
    func publishMlsMessage(_ payload: String, onError: @escaping (Error) -> Void = { _ in }) {
        guard !payload.isEmpty, payload.utf8.count <= 65_536 else {
            onError(SfuRoomAuthorizationError.invalidBinding)
            return
        }
        mls.addDocument(data: [
            "from": selfUid,
            "payload": payload,
            "createdAt": FieldValue.serverTimestamp(),
            "expireAt": expireAt()
        ], completion: { error in
            if let error { onError(error) }
        })
    }

    /// Subscribe to peers' MLS messages, INCLUDING the backlog present at attach — the peer can
    /// publish its KeyPackage before our listener attaches (the creator-claim transaction takes a
    /// round-trip first) and dropping it would deadlock the handshake. Batches are delivered in
    /// createdAt order (handshake messages are order-sensitive).
    func listenMlsMessages(
        onMessage: @escaping (_ payload: String, _ fromUid: String) -> Void,
        onError: @escaping (Error) -> Void = { _ in }
    ) -> ListenerRegistration {
        var seen = Set<String>()
        return mls.addSnapshotListener { [selfUid] snapshot, error in
            if let error { onError(error); return }
            guard let snapshot else { return }
            snapshot.documentChanges
                .filter { $0.type == .added && seen.insert($0.document.documentID).inserted }
                .map(\.document)
                .filter { ($0.get("from") as? String) != selfUid }
                .sorted {
                    let lhs = ($0.get("createdAt") as? Timestamp)?.dateValue() ?? .distantFuture
                    let rhs = ($1.get("createdAt") as? Timestamp)?.dateValue() ?? .distantFuture
                    return lhs < rhs
                }
                .forEach { document in
                    if let payload = document.get("payload") as? String,
                       let fromUid = document.get("from") as? String {
                        onMessage(payload, fromUid)
                    }
                }
        }
    }
}

File: Android/app/src/main/java/com/auralis/crisisconnect/messaging/call/sfu/SfuRoomClient.kt
package com.auralis.crisisconnect.messaging.call.sfu

import com.google.firebase.firestore.DocumentChange
import com.google.firebase.firestore.FieldValue
import com.google.firebase.firestore.FirebaseFirestore
import com.google.firebase.firestore.DocumentSnapshot
import com.google.firebase.firestore.ListenerRegistration
import com.google.firebase.Timestamp
import java.util.Date
import java.util.UUID
import kotlinx.coroutines.tasks.await

enum class SfuRoomScopeType(val wireValue: String) { AGENCY("agency"), HIERARCHY("hierarchy") }

/** Immutable v2 authorization record. Legacy/unbound rooms are never accepted. */
data class SfuRoomBinding(
    val scopeType: SfuRoomScopeType,
    val channelId: String,
    val callId: String,
    val participants: List<String>,
) {
    init {
        require(channelId.isNotBlank() && channelId.length <= 256)
        require(runCatching { UUID.fromString(callId) }.isSuccess)
        require(participants.size == 2 && participants[0] != participants[1])
        require(participants.all { it.isNotBlank() && it.length <= 128 })
    }

    val documentFields: Map<String, Any> get() = mapOf(
        "version" to 2,
        "scopeType" to scopeType.wireValue,
        "channelId" to channelId,
        "callId" to callId,
        "participants" to participants,
    )

    fun matches(snapshot: DocumentSnapshot): Boolean =
        snapshot.getLong("version") == 2L &&
            snapshot.getString("scopeType") == scopeType.wireValue &&
            snapshot.getString("channelId") == channelId &&
            snapshot.getString("callId") == callId &&
            (snapshot.get("participants") as? List<*>) == participants &&
            snapshot.getString("mlsCreator") in participants

    companion object {
        fun create(
            scopeType: SfuRoomScopeType,
            channelId: String,
            callId: String,
            selfUid: String,
            peerUid: String,
        ) = SfuRoomBinding(
            scopeType,
            channelId.trim(),
            callId.trim(),
            listOf(selfUid.trim(), peerUid.trim()).sorted(),
        )
    }
}

/**
 * SFU room roster over Firestore — the coordination plane the SFU itself doesn't provide. Mirrors the
 * web `sfu-room.ts`: `sfuRooms/{roomId}/participants/{uid}` holds each member's SFU sessionId + published
 * track names so everyone can pull everyone else. Carries NO media and NO plaintext keys — MLS handshake
 * bytes (Faz C) ride the sibling `mlsMessages` subcollection as opaque blobs. Metadata-only, so the same
 * membership-gated rules the web uses already allow mobile accounts.
 */
class SfuRoomClient(
    private val roomId: String,
    private val selfUid: String,
    private val binding: SfuRoomBinding,
    private val protocolVersion: SfuProtocolVersion = SfuProtocolVersion.SECURE,
    private val firestore: FirebaseFirestore = FirebaseFirestore.getInstance(),
) {
    val enforcesBoundIdentity: Boolean get() = true
    private val roomCollection = "sfuRoomsV2"
    private val participants = firestore.collection(roomCollection).document(roomId).collection("participants")
    private val mls = firestore.collection(roomCollection).document(roomId).collection("mlsMessages")

    // sfuRooms are ephemeral (one per call) but their docs — the room meta, participant roster, and the
    // MLS handshake blobs — are never cleaned except a participant's own graceful-leave delete. An
    // `expireAt` field on every write lets a Firestore TTL policy sweep the orphans (a crashed peer's
    // roster doc, all mlsMessages, the room doc). Live calls keep extending it via heartbeat, so an
    // in-progress call is never swept.
    private fun expireAt(): Timestamp = Timestamp(Date(System.currentTimeMillis() + ROOM_TTL_MS))

    /** Announce (or update) my SFU identity so other members can pull my tracks and show my profile. */
    fun publishSelf(
        name: String,
        photoUrl: String?,
        cameraOn: Boolean,
        muted: Boolean,
        sessionId: String,
        tracks: List<SfuPublishedTrack>,
        onError: (Exception) -> Unit = {},
    ) {
        val doc = hashMapOf(
            "uid" to selfUid,
            "name" to name,
            "photoUrl" to photoUrl,
            "cameraOn" to cameraOn,
            "muted" to muted,
            "sessionId" to sessionId,
            "tracks" to tracks.map { mapOf("trackName" to it.trackName, "kind" to it.kind, "source" to it.source) },
            "updatedAt" to FieldValue.serverTimestamp(),
            "expireAt" to expireAt(),
        )
        participants.document(selfUid).set(doc).addOnFailureListener(onError)
    }

    /** Remove myself from the room roster on hang-up (best-effort). */
    fun leave() {
        runCatching { participants.document(selfUid).delete() }
    }

    /**
     * Liveness heartbeat: refresh my roster entry's `updatedAt` so peers can tell a live member from a
     * crashed/killed one (an ungraceful exit never deletes the doc — only a stale timestamp reveals it).
     */
    fun heartbeat(onError: (Exception) -> Unit = {}) {
        participants.document(selfUid)
            .update(mapOf("updatedAt" to FieldValue.serverTimestamp(), "expireAt" to expireAt()))
            .addOnFailureListener(onError)
    }

    /**
     * Atomically claim the MLS group-creator role on the room doc (mirrors web `claimMlsCreator`).
     * Secure rules cannot authorize a transaction read of a missing, not-yet-bound room. Both peers
     * therefore race an immutable create; the loser verifies the winner's exact binding with a read.
     */
    suspend fun claimMlsCreator(): Boolean {
        require(protocolVersion == SfuProtocolVersion.SECURE)
        require(selfUid in binding.participants)
        val ref = firestore.collection(roomCollection).document(roomId)
        val fields = HashMap<String, Any>()
        fields.putAll(binding.documentFields)
        fields["mlsCreator"] = selfUid
        fields["createdAt"] = FieldValue.serverTimestamp()
        fields["expireAt"] = expireAt()
        return try {
            ref.set(fields).await()
            true
        } catch (createError: Exception) {
            val snapshot = runCatching { ref.get().await() }.getOrNull()
            if (snapshot == null || !snapshot.exists() || !binding.matches(snapshot)) throw createError
            val creator = snapshot.getString("mlsCreator")
            if (creator.isNullOrBlank() || creator !in binding.participants) throw createError
            creator == selfUid
        }
    }

    /** Subscribe to the room roster; delivers everyone except me. Returns the registration to remove. */
    fun listenRoster(
        onRoster: (List<SfuRemoteParticipant>) -> Unit,
        onError: (Exception) -> Unit = {},
    ): ListenerRegistration =
        participants.addSnapshotListener { snap, error ->
            if (error != null) { onError(error); return@addSnapshotListener }
            if (snap == null) return@addSnapshotListener
            val remotes = ArrayList<SfuRemoteParticipant>()
            for (d in snap.documents) {
                if (d.id == selfUid) continue
                val sessionId = d.getString("sessionId") ?: continue
                @Suppress("UNCHECKED_CAST")
                val rawTracks = d.get("tracks") as? List<Map<String, Any?>> ?: continue
                val tracks = rawTracks.mapNotNull { t ->
                    val trackName = t["trackName"] as? String ?: return@mapNotNull null
                    SfuPublishedTrack(
                        trackName = trackName,
                        kind = t["kind"] as? String ?: "audio",
                        source = t["source"] as? String ?: "mic",
                    )
                }
                remotes.add(
                    SfuRemoteParticipant(
                        uid = d.id,
                        sessionId = sessionId,
                        tracks = tracks,
                        updatedAtMillis = d.getTimestamp("updatedAt")?.toDate()?.time,
                    )
                )
            }
            onRoster(remotes)
        }

    // ---- MLS handshake relay (Faz C). Opaque blobs; the SFU + Firestore never see keys/plaintext. ----

    /** Broadcast one opaque MLS handshake message (Base64) to the room. */
    fun publishMlsMessage(payloadBase64: String, onError: (Exception) -> Unit = {}) {
        if (payloadBase64.isEmpty() || payloadBase64.length > MAX_MLS_PAYLOAD_CHARS) {
            onError(IllegalArgumentException("MLS relay payload is outside the allowed size."))
            return
        }
        mls.add(hashMapOf("from" to selfUid, "payload" to payloadBase64, "createdAt" to FieldValue.serverTimestamp(), "expireAt" to expireAt()))
            .addOnFailureListener(onError)
    }

    /**
     * Subscribe to peers' MLS messages, INCLUDING any backlog present at attach. Room ids are ephemeral
     * (fresh UUID per call), so everything in this collection belongs to this call's handshake — and the
     * peer can publish its KeyPackage before our listener attaches (our creator-claim transaction takes a
     * network round-trip first), which a skip-the-backlog listener would drop, deadlocking the handshake.
     * Batches are processed in createdAt order (handshake messages are order-sensitive); server timestamps
     * can be momentarily null on latency-compensated local echoes, but we never receive our own messages
     * here, so every doc we process has a resolved timestamp.
     */
    fun listenMlsMessages(
        onMessage: (payloadBase64: String, fromUid: String) -> Unit,
        onError: (Exception) -> Unit = {},
    ): ListenerRegistration {
        val seen = HashSet<String>()
        return mls.addSnapshotListener { snap, error ->
            if (error != null) { onError(error); return@addSnapshotListener }
            if (snap == null) return@addSnapshotListener
            snap.documentChanges
                .filter { it.type == DocumentChange.Type.ADDED && seen.add(it.document.id) }
                .map { it.document }
                .filter { it.getString("from") != selfUid }
                .sortedBy { it.getTimestamp("createdAt")?.toDate()?.time ?: Long.MAX_VALUE }
                .forEach { d ->
                    val payload = d.getString("payload")
                    val fromUid = d.getString("from")
                    if (payload != null && fromUid != null) onMessage(payload, fromUid)
                }
        }
    }

    private companion object {
        // 12h: comfortably longer than any real authority call, refreshed by heartbeat while live.
        private const val ROOM_TTL_MS = 12L * 60 * 60 * 1000
        private const val MAX_MLS_PAYLOAD_CHARS = 65_536
    }
}
